temporalcli/commands.schedule.go

```diff
@@ -0,0 +1,484 @@
+package temporalcli
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/temporalio/cli/temporalcli/internal/printer"
+	enumspb "go.temporal.io/api/enums/v1"
+	schedpb "go.temporal.io/api/schedule/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+type printableSchedule struct {
+	ScheduleId       any
+	CalendarSpecs    []any `cli:",cardOmitEmpty"`
+	IntervalSpecs    []any `cli:",cardOmitEmpty"`
+	WorkflowId       any
+	WorkflowType     any
+	Paused           any
+	Notes            any `cli:",cardOmitEmpty"`
+	NextRunTime      any
+	LastRunTime      any
+	RunningWorkflows []any
+	SearchAttributes any `cli:",cardOmitEmpty"`
+	Memo             any `cli:",cardOmitEmpty"`
+}
+
+func describeResultToPrintable(id string, desc *client.ScheduleDescription) *printableSchedule {
+	// TODO: should we include any other fields here, e.g. jitter, time zone, start/end time
+	out := &printableSchedule{
+		ScheduleId:       id,
+		Paused:           desc.Schedule.State.Paused,
+		Notes:            desc.Schedule.State.Note,
+		SearchAttributes: desc.SearchAttributes,
+		Memo:             desc.Memo,
+	}
+	specToPrintable(out, desc.Schedule.Spec)
+	if workflowAction, ok := desc.Schedule.Action.(*client.ScheduleWorkflowAction); ok {
+		out.WorkflowId = workflowAction.ID
+		out.WorkflowType = workflowAction.Workflow
+	}
+	if len(desc.Info.NextActionTimes) > 0 {
+		out.NextRunTime = desc.Info.NextActionTimes[0]
+	}
+	if l := len(desc.Info.RecentActions); l > 0 {
+		last := desc.Info.RecentActions[l-1]
+		out.LastRunTime = last.ScheduleTime
+	}
+	for _, w := range desc.Info.RunningWorkflows {
+		out.RunningWorkflows = append(out.RunningWorkflows, w.WorkflowID)
+	}
+	return out
+}
+
+func listEntryToPrintable(ent *client.ScheduleListEntry) *printableSchedule {
+	out := &printableSchedule{
+		ScheduleId:       ent.ID,
+		Paused:           ent.Paused,
+		Notes:            ent.Note,
+		WorkflowType:     ent.WorkflowType.Name,
+		SearchAttributes: ent.SearchAttributes,
+		Memo:             ent.Memo,
+	}
+	specToPrintable(out, ent.Spec)
+	if len(ent.NextActionTimes) > 0 {
+		out.NextRunTime = ent.NextActionTimes[0]
+	}
+	if l := len(ent.RecentActions); l > 0 {
+		last := ent.RecentActions[l-1]
+		out.LastRunTime = last.ScheduleTime
+	}
+	return out
+}
+
+func specToPrintable(out *printableSchedule, spec *client.ScheduleSpec) {
+	for _, cal := range spec.Calendars {
+		// TODO: This is not quite right for either json or text:
+		// For json: this doesn't come out as proper protojson, the field names are
+		// day_of_month instead of dayOfMonth.
```

Comment:
This is the cost of mixing proto and non-proto JSON but there's not an east solution. What we do in some other commands is separate the JSON representation structs from the text ones, and for JSON, make this be something like `[]json.Message` instead where we pre-JSON it. It became too confusing for complicated models to try to satisfy JSON and non-JSON needs with the same model (but if you must reuse models, consider more explicit JSON models then that meet the JSON field name needs).